--- openssl-1.0.1j.orig/apps/openssl.c	2014-10-15 12:53:39.000000000 +0000
+++ openssl-1.0.1j/apps/openssl.c	2014-12-15 16:48:14.463765573 +0000
@@ -345,8 +345,10 @@ int main(int Argc, char *ARGV[])
 		if (ERR_GET_REASON(ERR_peek_last_error())
 		    == CONF_R_NO_SUCH_FILE)
 			{
+#if 0 /* ANDROID */
 			BIO_printf(bio_err,
 				   "WARNING: can't open config file: %s\n",p);
+#endif
 			ERR_clear_error();
 			NCONF_free(config);
 			config = NULL;
